File: Java/src/Palindrome Partitioning II.java
public class Solution
{
  public int minCut(String s)
  {
    int length = s.length();
    int[] dp = new int[length + 1];
    boolean[][] matrix = new boolean[length + 1][length + 1];
    for (int i = 1; i <= length; i++)
    {
      dp[i] = i;
      for (int j = 1; j <= i; j++)
      {
        if (s.charAt(i - 1) == s.charAt(j - 1))
        {
          if (i - j < 2 || matrix[j + 1][i - 1])
          {
            matrix[j][i] = true;
            if (j == 1)
              dp[i] = 0;
            else
            {
              dp[i] = Math.min(dp[i], dp[j - 1] + 1);
            }
          }
        }
      }
    }
    return dp[length];
  }
}

File: Python/Palindrome Partitioning II.py
class Solution:
    # @param s, a string
    # @return an integer
    def minCut(self, s):
        dp = [0] * (len(s)+1)
        matrix = [[False for i in range(len(s)+1)] for j in range(len(s)+1)]
        for i in range(1, len(s) + 1):
            dp[i] = i
            for j in range(1, i+1):
                if s[i-1] == s[j-1]:
                    if i-j<2 or matrix[j+1][i-1]:
                        matrix[j][i] = True
                        if j==1: 
                            dp[i] = 0
                        else:
                            dp[i] = min(dp[i],dp[j-1]+1)
        return dp[len(s)]
        
s = Solution()
print s.minCut("aac")        